eth-rpc/server/servertypes.nim

```diff
@@ -13,15 +14,145 @@ type
     procs: newTable[string, RpcProc]()
   )
 
-proc register*(server: RpcServer, name: string, rpc: RpcProc) =
-  server.procs[name] = rpc
+var sharedServer: RpcServer
 
-proc unRegisterAll*(server: RpcServer) = server.procs.clear
+proc sharedRpcServer*(): RpcServer =
+  if sharedServer.isNil: sharedServer = newRpcServer("")
+  result = sharedServer
+  
+proc fromJson(n: JsonNode, argName: string, result: var bool) =
+  if n.kind != JBool: raise newException(ValueError, "Parameter \"" & argName & "\" expected JBool but got " & $n.kind)
+  result = n.getBool()
+
+proc fromJson(n: JsonNode, argName: string, result: var int) =
+  if n.kind != JInt: raise newException(ValueError, "Parameter \"" & argName & "\" expected JInt but got " & $n.kind)
+  result = n.getInt()
+
+proc fromJson(n: JsonNode, argName: string, result: var byte) =
+  if n.kind != JInt: raise newException(ValueError, "Parameter \"" & argName & "\" expected JInt but got " & $n.kind)
+  let v = n.getInt()
+  if v > 255 or v < 0: raise newException(ValueError, "Parameter \"" & argName & "\" value out of range for byte: " & $v)
+  result = byte(v)
+
+proc fromJson(n: JsonNode, argName: string, result: var float) =
+  if n.kind != JFloat: raise newException(ValueError, "Parameter \"" & argName & "\" expected JFloat but got " & $n.kind)
+  result = n.getFloat()
+
+proc fromJson(n: JsonNode, argName: string, result: var string) =
+  if n.kind != JString: raise newException(ValueError, "Parameter \"" & argName & "\" expected JString but got " & $n.kind)
+  result = n.getStr()
+
+proc fromJson[T](n: JsonNode, argName: string, result: var seq[T]) =
+  if n.kind != JArray: raise newException(ValueError, "Parameter \"" & argName & "\" expected JArray but got " & $n.kind)
+  result = newSeq[T](n.len)
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[N, T](n: JsonNode, argName: string, result: var array[N, T]) =
+  if n.kind != JArray: raise newException(ValueError, "Parameter \"" & argName & "\" expected JArray but got " & $n.kind)
+  if n.len > result.len: raise newException(ValueError, "Parameter \"" & argName & "\" item count is too big for array")
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[T: object](n: JsonNode, argName: string, result: var T) =
+  if n.kind != JObject: raise newException(ValueError, "Parameter \"" & argName & "\" expected JObject but got " & $n.kind)
+  for k, v in fieldpairs(result):
+    fromJson(n[k], k, v)
+
+proc unpackArg[T](argIdx: int, argName: string, argtype: typedesc[T], args: JsonNode): T =
+  fromJson(args[argIdx], argName, result)
+
+proc setupParams(parameters, paramsIdent: NimNode): NimNode =
+  # Add code to verify input and load parameters into Nim types
+  result = newStmtList()
+  if not parameters.isNil:
+    # initial parameter array length check
+    var expectedLen = parameters.len - 1
+    let expectedStr = "Expected " & $expectedLen & " Json parameter(s) but got "
```

Comment:
I'd better use a callout to some `validateParameters(expectedCount, jNode)`. This simplifies the macro, and keeps bin size smaller.